src/metabase/query_processor/middleware/desugar.clj

```diff
@@ -70,24 +71,20 @@
   `<unit>` is inferred from the `:datetime-field` the clause is being compared to (if any), otherwise falls back to
   `default.`"
   [query]
-  (mbql.u/replace-in query [:query :filter]
+  (mbql.u/replace query
     [clause field [:relative-datetime :current & _]]
     [clause field [:relative-datetime 0 (or (mbql.u/match-one field [:datetime-field _ unit] unit)
                                             :default)]]))
 
-
 (s/defn ^:private desugar* :- mbql.s/Query
-  [{{filter-clause :filter} :query, :as query}]
-  (if-not (seq filter-clause)
-    query
-    (-> query
-        desugar-inside
-        desugar-is-null-and-not-null
-        desugar-time-interval
-        desugar-does-not-contain
-        desugar-equals-and-not-equals-with-extra-args
-        desugar-current-relative-datetime
-        (update-in [:query :filter] mbql.u/simplify-compound-filter))))
+  [query]
+  (update query :query (comp mbql.u/simplify-compound-filter
+                             desugar-inside
```

Comment:
Walking the query dictionary 7 times is probably not going to be a deal breaker, because they're so small to begin with. When we spend 5 ms processing a query and 10 seconds running it I wouldn't worry too much about optimizing the 5 ms